### PR TITLE
T13 - Password

### PR DESCRIPTION
**Pull request by Priyanka15-12**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -1,7 +1,40 @@
...
+   internal class Program {
+      static void Main (string[] args) {
+         Console.Write ("Enter Password: ");
+         string Password = Console.ReadLine ();
+         if (check (Password) == "true") Console.WriteLine ("Strong password.");
+         else Console.WriteLine ("Not a strong password. \n" + check (Password)) ;
+      }
+      static string check (string pass) {
+         string lower = "", upper = "", digit = "", splchar = "", notstrong = "";
+         foreach (char c in pass) {
```
Comment (MetaNaveen) [inline Program.cs:11]: One loop is sufficient. Can you modify it accordingly.

File: Program.cs [hunk 2]
```diff
@@ -1,7 +1,40 @@
...
\ No newline at end of file
+   internal class Program {
+      static void Main (string[] args) {
+         Console.Write ("Enter Password: ");
+         string Password = Console.ReadLine ();
+         if (check (Password) == "true") Console.WriteLine ("Strong password.");
+         else Console.WriteLine ("Not a strong password. \n" + check (Password)) ;
+      }
+      static string check (string pass) {
+         string lower = "", upper = "", digit = "", splchar = "", notstrong = "";
```
Comment (MetaNaveen) [inline Program.cs:10]: Can you think of better **data type** rather than **string**?

File: Program.cs [hunk 3]
```diff
@@ -1,7 +1,40 @@
...
-            Console.WriteLine ("Hello, World!");
-        }
-    }
-}
\ No newline at end of file
+   internal class Program {
+      static void Main (string[] args) {
+         Console.Write ("Enter Password: ");
+         string Password = Console.ReadLine ();
+         if (check (Password) == "true") Console.WriteLine ("Strong password.");
```
Comment (MetaNaveen) [inline Program.cs:6]: Can you check for the data type [bool](https://www.w3schools.com/cs/cs_booleans.php)?

File: Program.cs [hunk 4]
```diff
@@ -1,7 +1,40 @@
...
+         foreach (char c in pass) {
+            if (char.IsUpper (c))
+               upper = "true";
+         }
+         foreach (char c in pass) {
+            if (char.IsDigit (c))
+               digit = "true";
+         }
+         foreach (char c in pass) {
+            string spl = " !@#$%^&*()-+";
```
Comment (MetaNaveen) [inline Program.cs:24]: Space is not a valid character for password.

File: Program.cs [hunk 5]
```diff
@@ -1,7 +1,40 @@
...
-}
\ No newline at end of file
+   internal class Program {
+      static void Main (string[] args) {
+         Console.Write ("Enter Password: ");
+         string Password = Console.ReadLine ();
+         if (check (Password) == "true") Console.WriteLine ("Strong password.");
+         else Console.WriteLine ("Not a strong password. \n" + check (Password)) ;
+      }
+      static string check (string pass) {
```
Comment (MetaNaveen) [inline Program.cs:9]: You could use something like this as return type. **(bool isStrong, string errorMsg)** 
With this, you can print error message if the isStrong property is false.

File: Program.cs [hunk 6]
```diff
@@ -1,7 +1,40 @@
...
+         }
+         if (pass.Length >= 6 && lower == "true" && upper == "true" && splchar == "true" && digit == "true")
+            return "true";
+         else {
+            if (upper != "true") notstrong = "Password must have atleast one upper character\n";
+            if (lower != "true") notstrong += "Password must have atleast one lower character\n";
+            if (digit != "true") notstrong += "Password must have atleast one digit \n";
+            if (splchar != "true") notstrong += "Password must have atleast one special character.\n";
+            if (pass.Length < 6) notstrong += "Password must have atleast six character.\n";
+            return notstrong;
```
Comment (MetaNaveen) [inline Program.cs:36]: I hope, stringbuilder can be used for composing list of errors.

**Review by MetaNaveen — CHANGES_REQUESTED** [review]

File: Program.cs (hunk 1 above)
Comment (Priyanka15-12) [reply Program.cs:15]: Code changed using one loop.

File: Program.cs (hunk 2 above)
Comment (Priyanka15-12) [reply Program.cs]: Instead of string used bool.

File: Program.cs (hunk 3 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed.

File: Program.cs (hunk 4 above)
Comment (Priyanka15-12) [reply Program.cs]: Space removed.

File: Program.cs (hunk 5 above)
Comment (Priyanka15-12) [reply Program.cs]: Changes are done.

File: Program.cs (hunk 6 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed.

File: Program.cs [hunk 7]
```diff
@@ -1,7 +1,34 @@
...
+         StringBuilder err = new ();
+         bool lower = false, upper = false, digit = false, splChar = false;
+         foreach (char c in pass) {
+            if (char.IsLower (c))
+               lower = true;
+            if (char.IsUpper (c))
+               upper = true;
+            if (char.IsDigit (c))
+               digit = true;
+            string spl = "!@#$%^&*()-+";
```
Comment (MetaNaveen) [inline Program.cs:22]: Repeated statement/assignment could be defined outside the scope.

File: Program.cs [hunk 8]
```diff
@@ -1,7 +1,34 @@
...
+         if (isStrong) Console.WriteLine ("Strong password.");
+         else Console.WriteLine ("Password is not strong.\n" + errorMsg);
+      }
+      static (bool isStrong, string errorMsg) checkPassword (string pass) {
+         StringBuilder err = new ();
+         bool lower = false, upper = false, digit = false, splChar = false;
+         foreach (char c in pass) {
+            if (char.IsLower (c))
+               lower = true;
+            if (char.IsUpper (c))
```
Comment (MetaNaveen) [inline Program.cs:18]: Do we need to check condition if the char c is already true for IsLower()? character satisfies only one of the conditions at a time.

File: Program.cs [hunk 9]
```diff
@@ -1,7 +1,34 @@
...
+         string Password = Console.ReadLine ();
+         (bool isStrong, string errorMsg) = checkPassword (Password);
+         if (isStrong) Console.WriteLine ("Strong password.");
+         else Console.WriteLine ("Password is not strong.\n" + errorMsg);
+      }
+      static (bool isStrong, string errorMsg) checkPassword (string pass) {
+         StringBuilder err = new ();
+         bool lower = false, upper = false, digit = false, splChar = false;
+         foreach (char c in pass) {
+            if (char.IsLower (c))
```
Comment (MetaNaveen) [inline Program.cs:16]: Can you write a custom method for either of IsLower, IsUpper or IsDigit method?

**Review by MetaNaveen — CHANGES_REQUESTED** [review]

File: Program.cs (hunk 7 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed

File: Program.cs (hunk 9 above)
Comment (Priyanka15-12) [reply Program.cs]: Changes are done.

File: Program.cs [hunk 10]
```diff
@@ -9,19 +9,24 @@ internal class Program {
          if (isStrong) Console.WriteLine ("Strong password.");
          else Console.WriteLine ("Password is not strong.\n" + errorMsg);
       }
+      static bool isLower (char c) {
```
Comment (MetaNaveen) [inline Program.cs:12]: Boundary condition error.

File: Program.cs [hunk 11]
```diff
@@ -9,19 +9,24 @@ internal class Program {
          if (isStrong) Console.WriteLine ("Strong password.");
          else Console.WriteLine ("Password is not strong.\n" + errorMsg);
       }
+      static bool isLower (char c) {
+         return (c > 'a' && c < 'z');
+      }
+      static bool isUpper (char c) {
+         return (c > 'A' && c < 'Z');
+      }
+      static bool isDigit (char c) {
```
Comment (MetaNaveen) [inline Program.cs:18]: Boundary condition error.

File: Program.cs [hunk 12]
```diff
@@ -9,19 +9,24 @@ internal class Program {
          if (isStrong) Console.WriteLine ("Strong password.");
          else Console.WriteLine ("Password is not strong.\n" + errorMsg);
       }
+      static bool isLower (char c) {
+         return (c > 'a' && c < 'z');
+      }
+      static bool isUpper (char c) {
```
Comment (MetaNaveen) [inline Program.cs:15]: Boundary condition error.

File: Program.cs [hunk 13]
```diff
@@ -9,19 +9,24 @@ internal class Program {
...
-               upper = true;
-            if (char.IsDigit (c))
-               digit = true;
-            string spl = "!@#$%^&*()-+";
-            if (spl.Contains (c))
-               splChar = true;
+            if (isLower (c)) lower = true;
+            else if (isUpper (c)) upper = true;
+            else if (isDigit (c)) digit = true;
+            else if (spl.Contains (c)) splChar = true;
```
Comment (MetaNaveen) [inline Program.cs:29]: Unsupported special characters should not be allowed. [In this example, tilde (~) or quotes (") should not be allowed]

**Review by MetaNaveen — CHANGES_REQUESTED** [review]

**Comment by MetaNaveen** [comment]
Add testcases and comments in the code before submission.

File: Program.cs (hunk 10 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed

File: Program.cs (hunk 13 above)
Comment (Priyanka15-12) [reply Program.cs]: Changes done in such a way  that if only valid characters are entered, the input is checked for strong password.

File: Program.cs (hunk 12 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed

File: Program.cs (hunk 11 above)
Comment (Priyanka15-12) [reply Program.cs]: Changed

**Review by MetaNaveen — CHANGES_REQUESTED** [review]
Make the fix.

File: Program.cs [hunk 14]
```diff
@@ -1,7 +1,72 @@
...
+      static bool IsDigit (char c) => (c >= '0' && c <= '9');
+
+      /// <summary>Check whether password is strong or not and finds the reasons for weak password</summary>
+      /// <param name="password">string password</param>
+      /// <returns>Its returns bool value true for strong password or false for not a strong password with string of error message</returns>
+      static (bool, string ) IsStrongPassword (string password) {
+         StringBuilder errorMsg = new ();
+         string spl = "!@#$%^&*()-+";
+         bool lower = false, upper = false, digit = false, splChar = false;
+         bool flag = password.All(c=> char.IsLetter (c) || char.IsDigit (c) || spl.Contains (c));
```
[the remark this answers is not in the document]
Comment (Priyanka15-12) [reply Program.cs]: Changes are done.

File: Program.cs [hunk 15]
```diff
@@ -1,7 +1,74 @@
...
+   /// <summary>Strong password</summary>
+   internal class Program {
+      #region Methods ------------------------------------------------
+      /// <summary>This method shows whether the password is strong and not, which is given by user</summary>
+      /// <param name="args">arguments</param>
+      static void Main (string[] args) {
+         Console.Write ("Enter Password: ");
+         string password = Console.ReadLine () ?? "";
+         (bool isStrong, string errorMsg) = IsStrongPassword (password);
+         Console.WriteLine (isStrong ? "Strong psssword" : $"Not a strong password \n{errorMsg}");
```
Comment (MetaNaveen) [inline Program.cs:27]: Correct the typo.

**Pull request closed without merging**